### ISSUE TITLE
Issues with the results

### ISSUE THREAD

**Issue by antonyvan** [pzins/Data-Driven-3D-Reconstruction-of-Dressed-Humans-from-Sparse-Views#5]
Hello, thank you so much for your sharing your work.
I generated data of 1500 objects and trained the network 100 epochs. the saved checkpoint is 
opt.load_checkpoint = "logs/ortho/version_3/checkpoints/epoch=97-step=823200-my_loss.ckpt"
I test the network and got the results like this  
![snapshot00](https://github.com/pzins/Data-Driven-3D-Reconstruction-of-Dressed-Humans-from-Sparse-Views/assets/40614822/2906fc0a-ff0b-46bf-987b-da5769f69dc1)

The only code I changed is line 153 which produce zeros if we convert sdf to np.float16
and line 158              # verts, faces, normals, values = measure.marching_cubes_lewiner(sdf, thres, gradient_direction=gdir) to              verts, faces, normals, values = measure.marching_cubes(sdf, thres)
because original marching_cubes_lewiner doesn't produce any verts.

could you help to take a look at it? 
Thanks a lot

**Comment by antonyvan** [comment]
follow up with my question,
I actually figure out one issue:  the ground truth mesh size should be in centimeter units. thus a human body height i.e. y values will be ~ 0 ---- 170.00.  Thus, with this data, the select_sampling_method in  data loader will generate roughly similar number of inside_points and outside points. 
Previous mesh scale is in meter, thus almost no inside_points. 

I think the data is correct now.  However the training is still not right.
I checked the code. and want to ask if the following are the problem: 

![Screenshot from 2024-01-08 11-48-06](https://github.com/pzins/Data-Driven-3D-Reconstruction-of-Dressed-Humans-from-Sparse-Views/assets/40614822/5cb25d92-6829-41b8-b530-44b200c334a6)

 in the figure, the xy in each camera has unit of pixels. thus the xy is around  minus hundreds ---- plus hundreds.
should we divide it by half of image size to normalize to scale to [-1, 1 ] so that line 172 is valid. ?

**Comment by pzins** [comment]
Hello,

1) yes, I used human meshes that are in centimeters. 
It's also possible to use meshes in meters but in that case, the training sample points and the evaluation grid at inference should also be in meters.


2) xy needs to be normalized in [-1,1] because they are used in torch.nn.functional.grid_sample() (utils.py project() function) to associate each sample point with the corresponding features of each view.
It seems that you are using orthographic projection, so if I remember correctly, the normalization of xy in [-1,1] is already done in the projection, so you don't need to do it on line 170.
It's possible that some xy points are not in [-1,1]. It just means that the corresponding 3D points are not visible in the image (they are projected out of the image) and the associated feature will be 0 (default padding_mode of torch.nn.functional.grid_sample())

**Comment by antonyvan** [comment]
Thanks a lot for your time. 
yes, I am using the orthographic projection.  Could you confirm that the xy is already normalized in the projection ?  Could you point out where the normalization is done?   The reason why I think it is not normalized is that once the query points are  normalized, mostly are within range [-1, 1]. but you see , in the figure, mostly are still outside of the range.  also without dividing 256 (half image size) the training doesn't looks right. the loss always stays at 0.5.
  
Thanks a lot :) 

By the way, do I need just simply set  grid_res = 3 in the configuration files in order to use your designed fusion ?  Current setup in the configuration file doesn't allow padding neighbor points for query point. so it looks like the current configuration is for multiview-PIFU. 

Thanks a lot

**Comment by pzins** [comment]
xy is already normalized in the projection (load_data_ortho() in BaseDataset.py) with the "uv_intrinsic" matrix. 
To debug you problem, maybe you can just sample many 3D points on the surface of your mesh (remove the random normal displacement with sigma = 0 in select_sampling_method()) and then project the points on each view, and verify that the result images are correct. For that, you can uncomment  "# self.debug_projections(xy)" on line 198 of MVPIFuNet.py.


I think I removed the support for grid_res > 1 when I cleaned the repo before making it public. The key element to improve the results compared to multiview-pifu is the view fusion module based on attention. Using local grids also improves the results but less significantly and  at the same time increases the training time and the memorry consumption. So I just removed it from this repo.

**Comment by antonyvan** [comment]
Thanks a lot for your explanation.  I finally figure out that it is still my data issue. some parts of my data preparation  use centimeter, some use meter. 
I see that you have the scaling in load_data_ortho of BaseDataset.py   and I have checked that using your debug_projections(xy) it is correct now. The training looks correct now.
  
![1-img](https://github.com/pzins/Data-Driven-3D-Reconstruction-of-Dressed-Humans-from-Sparse-Views/assets/40614822/4392f388-436d-4e2a-8202-3e87abe35c55)

For the transformer based fusion. I see, we probably don't need so many neighbor points of query points since the performance boots vs memory consumption is not that good.
Thanks a lot for your detailed tutorial.  Really appreciated ! :)
